### DISCUSSION TITLE
FadeQuad

### DISCUSSION THREAD

**Discussion by night2876** [xoreos/xoreos#588] [General]
Hi,
i'm continuing to develop NWN2 code using Xoreos as base. I'm developing all the GUI/UI without problem, but i don't understand how made a fadein, fadeout between GUI.
I found in graphics/aurora FadeQuad, but i don't understand how to use it.
Must create a FadeQuad object, set the wait,run,start arguments and add in the show() of the GUI ?
Someone could help me understand how this work ?

Best Regards

**Comment by DrMcCoy** [comment]
Have a look how it's used in `src/engines/kotorbase`:

- Create the FadeQuad object (for example with `std::unique_ptr<Graphics::Aurora::FadeQuad> _fade;` as a member variable and `_fade = std::make_unique<Graphics::Aurora::FadeQuad>();` in the constructor).
- `_fade->setColor(r, g, b)` to set the blank color
- `_fade->setWaitTime(wait)` to set the wait time, as a float, in seconds. Can be 0.0f, of course
- `_fade->setRunTime(run)`  to set the fading time, as a float, in seconds
- Then you call either `_fade->fadeOut()` or `_fade->fadeIn()` to start fading out or in, respectively

As soon as you call either `fadeOut()` or `fadeIn()`, the wait time starts to run. As soon as the wait time is over, fading starts.

Important: you can't just create the FadeQuad as a function-local variable, set the parameters, start the fade and then end the function, because that will destroy the FadeQuad at the end of the function. You need to keep the FadeQuad around, for example as a member variable in a class.

**Reply by night2876** [reply]
Found references only in module.h and module.cpp in the kotorbase dir. But is not used only create the object.
The other reference is in a functions_module.cpp ...

**Comment by DrMcCoy** [comment]
Yes, functions_module.cpp uses it. The FadeQuad is stored in the Module, so module.cpp creates it and function_module.cpp access it.  In the KotOR games, it's used for scripting. I.e. the script can say fade in or out, and the FadeQuad will do that.

**Reply by night2876** [reply]
Example ...

I'm in the mainMenu creation

void Game::mainMenu(bool playStartSound, bool showLegal) {
	playMenuMusic();

	if (playStartSound)
		playSound("gui_prompt", Sound::kSoundTypeSFX);

	EventMan.flushEvents();

	MainMenu menu(*_campaign, _console);

	menu.show();
        
	menu.run();
        
	menu.hide();

	stopMenuMusic();
}

how add a Fade in beetween the show() and the run()

i've tryied to add this code but nothing.

        std::unique_ptr<Graphics::Aurora::FadeQuad> _fade = std::make_unique<Graphics::Aurora::FadeQuad>();
        _fade->setColor(1, 1, 1);
	_fade->setWaitTime(0.0f);
	_fade->setRunTime(10.0f);
	_fade->fadeIn();

**Reply by night2876** [reply]
Hi,

solved the problem. Was more complex than use the FadeQuad because FadeQuad belong on renderable and not in the kQueueVisibleGUIFrontObject so when rendering on the GUI queue the OpenGL don't process the request at all ... I created a new Class guifade to use in the gui queue events manager.

Thanks anyway.
